File: test/glue/app-model/host_wasm.c
#include "roc_platform_abi.h"

#include <stddef.h>
#include <stdint.h>

enum {
    MAX_ALLOCATIONS = 512,
    CANARY_SIZE = 16,
    CANARY_BYTE = 0xA5,
    POISON_BYTE = 0xCC,
    WASM_PAGE_SIZE = 65536,
};

#define WASM_EXPORT(name) __attribute__((export_name(name))) __attribute__((used))

typedef struct {
    uint8_t *raw;
    uint8_t *user;
    size_t length;
    size_t total;
    size_t alignment;
    int live;
} Allocation;

static Allocation allocations[MAX_ALLOCATIONS];
static uintptr_t heap_cursor = 0;
static size_t alloc_count = 0;
static size_t dealloc_count = 0;
static size_t live_alloc_count = 0;
static size_t allocator_error_count = 0;
static size_t failure_count = 0;
static char report[1024];
static size_t report_len = 0;

static size_t cstr_len(const char *bytes) {
    size_t len = 0;
    while (bytes[len] != 0) {
        len += 1;
    }
    return len;
}

static void memory_set(uint8_t *dst, uint8_t value, size_t len) {
    for (size_t i = 0; i < len; i += 1) {
        dst[i] = value;
    }
}

static void memory_copy(void *dst_void, const void *src_void, size_t len) {
    uint8_t *dst = (uint8_t *)dst_void;
    const uint8_t *src = (const uint8_t *)src_void;
    for (size_t i = 0; i < len; i += 1) {
        dst[i] = src[i];
    }
}

static int memory_equal(const void *left_void, const void *right_void, size_t len) {
    const uint8_t *left = (const uint8_t *)left_void;
    const uint8_t *right = (const uint8_t *)right_void;
    for (size_t i = 0; i < len; i += 1) {
        if (left[i] != right[i]) {
            return 0;
        }
    }
    return 1;
}

static int is_power_of_two(size_t value) {
    return value != 0 && (value & (value - 1u)) == 0;
}

static uintptr_t align_forward(uintptr_t value, size_t alignment) {
    return (value + alignment - 1u) & ~(uintptr_t)(alignment - 1u);
}

static void set_report(const char *prefix, const char *message) {
    if (failure_count != 0) {
        return;
    }

    const size_t prefix_len = cstr_len(prefix);
    const size_t message_len = cstr_len(message);
    size_t written = 0;
    while (written < prefix_len && written + 1u < sizeof(report)) {
        report[written] = prefix[written];
        written += 1;
    }
    for (size_t i = 0; i < message_len && written + 1u < sizeof(report); i += 1) {
        report[written] = message[i];
        written += 1;
    }
    report[written] = 0;
    report_len = written;
}

static void record_failure(const char *message) {
    set_report("FAIL app-model CGlue wasm32: ", message);
    failure_count += 1;
}

static void record_allocator_failure(const char *message) {
    allocator_error_count += 1;
    set_report("FAIL app-model CGlue wasm32 allocator: ", message);
    failure_count += 1;
}

static int ensure_wasm_memory(uintptr_t end) {
    const size_t required_pages = ((size_t)end + (size_t)WASM_PAGE_SIZE - 1u) / (size_t)WASM_PAGE_SIZE;
    const size_t current_pages = __builtin_wasm_memory_size(0);
    if (required_pages <= current_pages) {
        return 1;
    }
    const size_t previous_pages = __builtin_wasm_memory_grow(0, required_pages - current_pages);
    return previous_pages != (size_t)-1;
}

static void *bump_alloc(size_t total, size_t alignment) {
    if (heap_cursor == 0) {
        heap_cursor = (uintptr_t)(__builtin_wasm_memory_size(0) * WASM_PAGE_SIZE);
    }
    const uintptr_t raw = align_forward(heap_cursor, alignment);
    if (raw > UINTPTR_MAX - total) {
        record_allocator_failure("bump allocation overflow");
        return NULL;
    }
    const uintptr_t end = raw + total;
    if (!ensure_wasm_memory(end)) {
        record_allocator_failure("wasm memory grow failed");
        return NULL;
    }
    heap_cursor = end;
    return (void *)raw;
}

static Allocation *find_allocation(void *ptr) {
    for (size_t i = 0; i < MAX_ALLOCATIONS; i += 1) {
        if (allocations[i].live && allocations[i].user == (uint8_t *)ptr) {
            return &allocations[i];
        }
    }
    return NULL;
}

static int check_canaries(const Allocation *allocation) {
    for (size_t i = 0; i < CANARY_SIZE; i += 1) {
        if (allocation->user[-(ptrdiff_t)CANARY_SIZE + (ptrdiff_t)i] != CANARY_BYTE) {
            record_allocator_failure("prefix canary changed");
            return 1;
        }
        if (allocation->user[allocation->length + i] != CANARY_BYTE) {
            record_allocator_failure("suffix canary changed");
            return 1;
        }
    }
    return 0;
}

static void *contract_alloc(size_t length, size_t alignment) {
    if (!is_power_of_two(alignment)) {
        record_allocator_failure("invalid alignment");
        return NULL;
    }
    if (length > SIZE_MAX - CANARY_SIZE - CANARY_SIZE - alignment) {
        record_allocator_failure("allocation size overflow");
        return NULL;
    }

    const size_t total = CANARY_SIZE + alignment - 1u + length + CANARY_SIZE;
    uint8_t *raw = (uint8_t *)bump_alloc(total == 0 ? 1 : total, alignment);
    if (raw == NULL) {
        return NULL;
    }

    uint8_t *user = (uint8_t *)align_forward((uintptr_t)(raw + CANARY_SIZE), alignment);
    if (((uintptr_t)user % alignment) != 0) {
        record_allocator_failure("returned pointer is not aligned");
        return NULL;
    }

    Allocation *slot = NULL;
    for (size_t i = 0; i < MAX_ALLOCATIONS; i += 1) {
        if (!allocations[i].live) {
            slot = &allocations[i];
            break;
        }
    }
    if (slot == NULL) {
        record_allocator_failure("allocation table exhausted");
        return NULL;
    }

    memory_set(user - CANARY_SIZE, CANARY_BYTE, CANARY_SIZE);
    memory_set(user, POISON_BYTE, length);
    memory_set(user + length, CANARY_BYTE, CANARY_SIZE);

    *slot = (Allocation){
        .raw = raw,
        .user = user,
        .length = length,
        .total = total,
        .alignment = alignment,
        .live = 1,
    };
    alloc_count += 1;
    live_alloc_count += 1;
    return user;
}

static void contract_dealloc(void *ptr, size_t alignment) {
    if (ptr == NULL) {
        return;
    }
    Allocation *allocation = find_allocation(ptr);
    if (allocation == NULL) {
        record_allocator_failure("unknown or double free");
        return;
    }
    if (allocation->alignment != alignment) {
        record_allocator_failure("dealloc alignment mismatch");
    }
    (void)check_canaries(allocation);
    memory_set(allocation->user, 0xDD, allocation->length);
    allocation->live = 0;
    dealloc_count += 1;
    live_alloc_count -= 1;
}

static void *contract_realloc(void *ptr, size_t new_length, size_t alignment) {
    if (ptr == NULL) {
        return contract_alloc(new_length, alignment);
    }
    Allocation *old = find_allocation(ptr);
    if (old == NULL) {
        record_allocator_failure("realloc unknown pointer");
        return NULL;
    }
    if (old->alignment != alignment) {
        record_allocator_failure("realloc alignment mismatch");
        return NULL;
    }
    if (check_canaries(old) != 0) {
        return NULL;
    }

    const size_t copy_length = old->length < new_length ? old->length : new_length;
    void *new_ptr = contract_alloc(new_length, alignment);
    if (new_ptr == NULL) {
        return NULL;
    }
    memory_copy(new_ptr, ptr, copy_length);
    if (!memory_equal(new_ptr, ptr, copy_length)) {
        record_allocator_failure("realloc did not preserve old bytes");
    }
    contract_dealloc(ptr, alignment);
    return new_ptr;
}

void *roc_alloc(size_t length, size_t alignment) {
    return contract_alloc(length, alignment);
}

void roc_dealloc(void *ptr, size_t alignment) {
    contract_dealloc(ptr, alignment);
}

void *roc_realloc(void *ptr, size_t new_length, size_t alignment) {
    return contract_realloc(ptr, new_length, alignment);
}

void roc_dbg(const uint8_t *bytes, size_t len) {
    (void)bytes;
    (void)len;
}

void roc_expect_failed(const uint8_t *bytes, size_t len) {
    (void)bytes;
    (void)len;
    record_failure("roc_expect_failed");
}

void roc_crashed(const uint8_t *bytes, size_t len) {
    (void)bytes;
    (void)len;
    record_failure("roc_crashed");
}

static int roc_str_is_small(RocStr str) {
    return (intptr_t)str.length < 0;
}

static size_t roc_str_len(const RocStr *str) {
    if (roc_str_is_small(*str)) {
        return ((const uint8_t *)str)[sizeof(RocStr) - 1u] ^ 0x80u;
    }
    return str->length;
}

static const uint8_t *roc_str_bytes(const RocStr *str) {
    if (roc_str_is_small(*str)) {
        return (const uint8_t *)str;
    }
    return str->bytes;
}

static uint8_t *roc_str_allocation_ptr(RocStr str) {
    if (roc_str_is_small(str)) {
        return NULL;
    }
    if ((str.capacity_or_alloc_ptr & 1u) != 0) {
        return (uint8_t *)(str.capacity_or_alloc_ptr & ~(uintptr_t)1u);
    }
    return str.bytes;
}

static void roc_str_decref(RocStr str) {
    uint8_t *alloc_ptr = roc_str_allocation_ptr(str);
    if (alloc_ptr == NULL) {
        return;
    }
    intptr_t *rc = (intptr_t *)(alloc_ptr - sizeof(intptr_t));
    if (*rc == 0) {
        return;
    }
    *rc -= 1;
    if (*rc == 0) {
        roc_dealloc(alloc_ptr - sizeof(size_t), ROC_ALIGNOF(size_t));
    }
}

enum {
#if UINTPTR_MAX == UINT64_MAX
    MSG_TAG_OFFSET = 24,
    VIEW_TITLE_OFFSET = 0,
    VIEW_MODEL_OFFSET = 24,
    VIEW_MESSAGES_OFFSET = 32,
    VIEW_LIFECYCLE_TAG_OFFSET = 88,
#else
    MSG_TAG_OFFSET = 12,
    VIEW_TITLE_OFFSET = 0,
    VIEW_MODEL_OFFSET = 12,
    VIEW_MESSAGES_OFFSET = 16,
    VIEW_LIFECYCLE_TAG_OFFSET = 44,
#endif
    MSG_RESET_TAG = 3,
    LIFECYCLE_READY_TAG = 1,
};

static Msg reset_msg(void) {
    Msg msg = {{0}};
    msg.bytes[MSG_TAG_OFFSET] = MSG_RESET_TAG;
    return msg;
}

static void roc_box_decref(RocBox box) {
    if (box == 0) {
        return;
    }
    uint8_t *data = (uint8_t *)box;
    intptr_t *rc = (intptr_t *)(data - sizeof(intptr_t));
    if (*rc == 0) {
        return;
    }
    *rc -= 1;
    if (*rc == 0) {
        roc_dealloc(data - sizeof(size_t), ROC_ALIGNOF(size_t));
    }
}

static RocList *view_messages(View *view) {
    return (RocList *)(void *)(view->bytes + VIEW_MESSAGES_OFFSET);
}

static RocBox *view_model(View *view) {
    return (RocBox *)(void *)(view->bytes + VIEW_MODEL_OFFSET);
}

static RocStr *view_title(View *view) {
    return (RocStr *)(void *)(view->bytes + VIEW_TITLE_OFFSET);
}

static void view_decref(View view) {
    RocStr *title = view_title(&view);
    roc_str_decref(*title);
    roc_box_decref(*view_model(&view));
}

static void reset_contract(void) {
    for (size_t i = 0; i < MAX_ALLOCATIONS; i += 1) {
        allocations[i] = (Allocation){0};
    }
    alloc_count = 0;
    dealloc_count = 0;
    live_alloc_count = 0;
    allocator_error_count = 0;
    failure_count = 0;
    report_len = 0;
    report[0] = 0;
}

static void finish_pass(void) {
    const char *message = "PASS glue-runtime app-model CGlue wasm32";
    report_len = cstr_len(message);
    memory_copy(report, message, report_len);
    report[report_len] = 0;
}

static void run_contract(void) {
    RocBox initial = roc_init();
    RocBox updated = roc_update(initial, reset_msg());
    View view = roc_render(updated);

    RocStr *title = view_title(&view);
    const char *expected_title = "ready";
    const size_t expected_len = cstr_len(expected_title);
    if (roc_str_len(title) != expected_len || !memory_equal(roc_str_bytes(title), expected_title, expected_len)) {
        record_failure("render title mismatch");
    }
    if (view.bytes[VIEW_LIFECYCLE_TAG_OFFSET] != LIFECYCLE_READY_TAG) {
        record_failure("render lifecycle expected Ready");
    }
    RocList *messages = view_messages(&view);
    if (messages->length != 0) {
        record_failure("render messages expected empty");
    }
    view_decref(view);

    if (allocator_error_count != 0) {
        record_failure("allocator recorded errors");
    }
    if (live_alloc_count != 0) {
        record_failure("live allocations after scenario");
    }

    if (failure_count == 0) {
        finish_pass();
    } else if (report_len == 0) {
        set_report("FAIL app-model CGlue wasm32: ", "unknown failure");
    }
}

WASM_EXPORT("wasm_main")
const uint8_t *wasm_main(void) {
    reset_contract();
    run_contract();
    return (const uint8_t *)report;
}

WASM_EXPORT("wasm_result_len")
size_t wasm_result_len(void) {
    return report_len;
}

WASM_EXPORT("wasm_alloc_count")
size_t wasm_alloc_count(void) {
    return alloc_count;
}

WASM_EXPORT("wasm_dealloc_count")
size_t wasm_dealloc_count(void) {
    return dealloc_count;
}

File: test/glue/app-model/host_wasm.zig
const abi = @import("roc_platform_abi.zig");

const max_allocations = 512;
const canary_size = 16;
const canary_byte: u8 = 0xA5;
const poison_byte: u8 = 0xCC;
const wasm_page_size = 65_536;

const Allocation = struct {
    user: ?[*]u8 = null,
    length: usize = 0,
    alignment: usize = 0,
    live: bool = false,
};

const ContractEnv = struct {
    allocations: [max_allocations]Allocation = [_]Allocation{.{}} ** max_allocations,
    heap_cursor: usize = 0,
    alloc_count: usize = 0,
    dealloc_count: usize = 0,
    live_alloc_count: usize = 0,
    allocator_error_count: usize = 0,
    failure_count: usize = 0,
    report: [1024]u8 = [_]u8{0} ** 1024,
    report_len: usize = 0,

    fn reset(self: *ContractEnv) void {
        self.* = .{};
    }

    fn setReport(self: *ContractEnv, comptime prefix: []const u8, comptime message: []const u8) void {
        if (self.failure_count != 0) return;
        const text = prefix ++ message;
        @memcpy(self.report[0..text.len], text);
        self.report_len = text.len;
    }

    fn fail(self: *ContractEnv, comptime message: []const u8) void {
        self.setReport("FAIL app-model wasm32: ", message);
        self.failure_count += 1;
    }

    fn allocatorFail(self: *ContractEnv, comptime message: []const u8) void {
        self.allocator_error_count += 1;
        self.setReport("FAIL app-model wasm32 allocator: ", message);
        self.failure_count += 1;
    }

    fn finishPass(self: *ContractEnv) void {
        const message = "PASS glue-runtime app-model ZigGlue wasm32";
        @memcpy(self.report[0..message.len], message);
        self.report_len = message.len;
    }

    fn findAllocation(self: *ContractEnv, ptr: *anyopaque) ?*Allocation {
        const needle = @intFromPtr(ptr);
        for (&self.allocations) |*allocation| {
            if (allocation.live and @intFromPtr(allocation.user.?) == needle) return allocation;
        }
        return null;
    }

    fn checkCanaries(self: *ContractEnv, allocation: *const Allocation) bool {
        const user = allocation.user.?;
        for (0..canary_size) |i| {
            if ((user - canary_size)[i] != canary_byte) {
                self.allocatorFail("prefix canary changed");
                return false;
            }
            if ((user + allocation.length)[i] != canary_byte) {
                self.allocatorFail("suffix canary changed");
                return false;
            }
        }
        return true;
    }

    fn bumpAlloc(self: *ContractEnv, total: usize, alignment: usize) ?[*]u8 {
        if (self.heap_cursor == 0) {
            const heap_start = @mulWithOverflow(@wasmMemorySize(0), wasm_page_size);
            if (heap_start[1] != 0) {
                self.allocatorFail("wasm memory exhausted");
                return null;
            }
            self.heap_cursor = heap_start[0];
        }
        const raw = alignForward(self.heap_cursor, alignment) orelse {
            self.allocatorFail("bump alignment overflow");
            return null;
        };
        const end_result = @addWithOverflow(raw, total);
        if (end_result[1] != 0) {
            self.allocatorFail("bump allocation overflow");
            return null;
        }
        const end = end_result[0];
        const required_pages = wasmPagesForBytes(end);
        const current_pages = @wasmMemorySize(0);
        if (required_pages > current_pages and @wasmMemoryGrow(0, required_pages - current_pages) == -1) {
            self.allocatorFail("wasm memory grow failed");
            return null;
        }
        self.heap_cursor = end;
        return @ptrFromInt(raw);
    }

    fn alloc(self: *ContractEnv, length: usize, alignment: usize) ?*anyopaque {
        if (alignment == 0 or (alignment & (alignment - 1)) != 0) {
            self.allocatorFail("invalid alignment");
            return null;
        }
        if (length > ~@as(usize, 0) - canary_size - canary_size - alignment) {
            self.allocatorFail("allocation size overflow");
            return null;
        }

        const total = canary_size + alignment - 1 + length + canary_size;
        const raw = self.bumpAlloc(@max(total, 1), alignment) orelse return null;
        const user_start = @addWithOverflow(@intFromPtr(raw), canary_size);
        if (user_start[1] != 0) {
            self.allocatorFail("user pointer overflow");
            return null;
        }
        const user_addr = alignForward(user_start[0], alignment) orelse {
            self.allocatorFail("user pointer alignment overflow");
            return null;
        };
        const user: [*]u8 = @ptrFromInt(user_addr);
        if (user_addr % alignment != 0) {
            self.allocatorFail("returned pointer is not aligned");
            return null;
        }

        var slot: ?*Allocation = null;
        for (&self.allocations) |*allocation| {
            if (!allocation.live) {
                slot = allocation;
                break;
            }
        }
        const allocation = slot orelse {
            self.allocatorFail("allocation table exhausted");
            return null;
        };

        @memset((user - canary_size)[0..canary_size], canary_byte);
        @memset(user[0..length], poison_byte);
        @memset((user + length)[0..canary_size], canary_byte);

        allocation.* = .{
            .user = user,
            .length = length,
            .alignment = alignment,
            .live = true,
        };
        self.alloc_count += 1;
        self.live_alloc_count += 1;
        return @ptrCast(user);
    }

    fn dealloc(self: *ContractEnv, ptr: ?*anyopaque, alignment: usize) void {
        const raw_ptr = ptr orelse return;
        const allocation = self.findAllocation(raw_ptr) orelse {
            self.allocatorFail("unknown or double free");
            return;
        };
        if (allocation.alignment != alignment) {
            self.allocatorFail("dealloc alignment mismatch");
        }
        _ = self.checkCanaries(allocation);
        @memset(allocation.user.?[0..allocation.length], 0xDD);
        allocation.live = false;
        self.dealloc_count += 1;
        self.live_alloc_count -= 1;
    }

    fn realloc(self: *ContractEnv, ptr: ?*anyopaque, new_length: usize, alignment: usize) ?*anyopaque {
        const raw_ptr = ptr orelse return self.alloc(new_length, alignment);
        const old = self.findAllocation(raw_ptr) orelse {
            self.allocatorFail("realloc unknown pointer");
            return null;
        };
        if (old.alignment != alignment) {
            self.allocatorFail("realloc alignment mismatch");
            return null;
        }
        if (!self.checkCanaries(old)) return null;

        const old_user = old.user.?;
        const copy_length = @min(old.length, new_length);
        const new_ptr = self.alloc(new_length, alignment) orelse return null;
        const new_user: [*]u8 = @ptrCast(new_ptr);
        @memcpy(new_user[0..copy_length], old_user[0..copy_length]);
        if (!bytesEqual(new_user[0..copy_length], old_user[0..copy_length])) {
            self.allocatorFail("realloc did not preserve old bytes");
        }
        self.dealloc(raw_ptr, alignment);
        return new_ptr;
    }
};

fn alignForward(value: usize, alignment: usize) ?usize {
    const sum = @addWithOverflow(value, alignment - 1);
    if (sum[1] != 0) return null;
    return sum[0] & ~(alignment - 1);
}

fn wasmPagesForBytes(byte_count: usize) usize {
    return byte_count / wasm_page_size + @intFromBool(byte_count % wasm_page_size != 0);
}

comptime {
    const max_usize = ~@as(usize, 0);
    if (wasmPagesForBytes(max_usize) != max_usize / wasm_page_size + 1) {
        @compileError("wasm page rounding must handle the usize limit");
    }
}

fn bytesEqual(a: []const u8, b: []const u8) bool {
    if (a.len != b.len) return false;
    for (a, b) |a_byte, b_byte| if (a_byte != b_byte) return false;
    return true;
}

var contract_env: ContractEnv = .{};
var roc_host = abi.RocHost{
    .env = @ptrCast(&contract_env),
    .roc_alloc = &hostAlloc,
    .roc_dealloc = &hostDealloc,
    .roc_realloc = &hostRealloc,
    .roc_dbg = &hostDbg,
    .roc_expect_failed = &hostExpectFailed,
    .roc_crashed = &hostCrashed,
};

fn hostAlloc(roc_host_ptr: *abi.RocHost, length: usize, alignment: usize) callconv(.c) ?*anyopaque {
    const env: *ContractEnv = @ptrCast(@alignCast(roc_host_ptr.env));
    return env.alloc(length, alignment);
}

fn hostDealloc(roc_host_ptr: *abi.RocHost, ptr: *anyopaque, alignment: usize) callconv(.c) void {
    const env: *ContractEnv = @ptrCast(@alignCast(roc_host_ptr.env));
    env.dealloc(ptr, alignment);
}

fn hostRealloc(roc_host_ptr: *abi.RocHost, ptr: *anyopaque, new_length: usize, alignment: usize) callconv(.c) ?*anyopaque {
    const env: *ContractEnv = @ptrCast(@alignCast(roc_host_ptr.env));
    return env.realloc(ptr, new_length, alignment);
}

fn hostDbg(_: *abi.RocHost, bytes: [*]const u8, len: usize) callconv(.c) void {
    _ = .{ bytes, len };
}

fn hostExpectFailed(roc_host_ptr: *abi.RocHost, bytes: [*]const u8, len: usize) callconv(.c) void {
    _ = .{ bytes, len };
    const env: *ContractEnv = @ptrCast(@alignCast(roc_host_ptr.env));
    env.fail("roc_expect_failed");
}

fn hostCrashed(roc_host_ptr: *abi.RocHost, bytes: [*]const u8, len: usize) callconv(.c) void {
    _ = .{ bytes, len };
    const env: *ContractEnv = @ptrCast(@alignCast(roc_host_ptr.env));
    env.fail("roc_crashed");
}

export fn roc_alloc(length: usize, alignment: usize) callconv(.c) ?*anyopaque {
    return contract_env.alloc(length, alignment);
}

export fn roc_dealloc(ptr: ?*anyopaque, alignment: usize) callconv(.c) void {
    contract_env.dealloc(ptr, alignment);
}

export fn roc_realloc(ptr: ?*anyopaque, new_length: usize, alignment: usize) callconv(.c) ?*anyopaque {
    return contract_env.realloc(ptr, new_length, alignment);
}

export fn roc_dbg(bytes: [*]const u8, len: usize) callconv(.c) void {
    _ = .{ bytes, len };
}

export fn roc_expect_failed(bytes: [*]const u8, len: usize) callconv(.c) void {
    _ = .{ bytes, len };
    contract_env.fail("roc_expect_failed");
}

export fn roc_crashed(bytes: [*]const u8, len: usize) callconv(.c) void {
    _ = .{ bytes, len };
    contract_env.fail("roc_crashed");
}

fn resetMsg() abi.Msg {
    var msg: abi.Msg = undefined;
    const bytes: [*]u8 = @ptrCast(&msg);
    @memset(bytes[0..@sizeOf(abi.Msg)], 0);
    msg.tag = .Reset;
    return msg;
}

fn rocStrSlice(str: *const abi.RocStr) ?[]const u8 {
    if (str.isSmallStr()) {
        const bytes: [*]const u8 = @ptrCast(str);
        return bytes[0..str.len()];
    }
    const bytes = str.bytes orelse return null;
    return bytes[0..str.len()];
}

fn runContract() void {
    const initial = abi.roc_init();
    const updated = abi.roc_update(initial, resetMsg());
    const view = abi.roc_render(updated);

    const title = view.title;
    const title_slice = rocStrSlice(&title) orelse {
        contract_env.fail("render title has null bytes");
        view.decref(&roc_host);
        return;
    };
    if (!bytesEqual(title_slice, "ready")) {
        contract_env.fail("render title mismatch");
    }
    if (view.lifecycle.tag != .Ready) {
        contract_env.fail("render lifecycle mismatch");
    }
    if (view.messages.length != 0) {
        contract_env.fail("render messages expected empty");
    }
    view.decref(&roc_host);

    if (contract_env.allocator_error_count != 0) {
        contract_env.fail("allocator recorded errors");
    }
    if (contract_env.live_alloc_count != 0) {
        contract_env.fail("live allocations after scenario");
    }

    if (contract_env.failure_count == 0) {
        contract_env.finishPass();
    } else if (contract_env.report_len == 0) {
        const unknown = "FAIL app-model wasm32: unknown failure";
        @memcpy(contract_env.report[0..unknown.len], unknown);
        contract_env.report_len = unknown.len;
    }
}

export fn wasm_main() [*]const u8 {
    contract_env.reset();
    runContract();
    return &contract_env.report;
}

export fn wasm_result_len() usize {
    return contract_env.report_len;
}

export fn wasm_alloc_count() usize {
    return contract_env.alloc_count;
}

export fn wasm_dealloc_count() usize {
    return contract_env.dealloc_count;
}
